### ISSUE TITLE
kfctl: handle CRD / Webhook establishment

### ISSUE THREAD

**Issue by yanniszark** [kubeflow/kubeflow#4327]
/kind bug

**What steps did you take and what happened:**
Kfctl will sometimes fail because of the following reasons:
1. CRDs haven't become established yet and CR creation fails.
2. Admission Webhooks haven't become ready yet and CR creation fails.

**What did you expect to happen:**
Kfctl should have some mechanism to counter these kind of scenarios.

**Anything else you would like to add:**
Some more details here: https://github.com/kubeflow/kubeflow/issues/3989#issuecomment-532638203
TLDR: the current way of defending against these scenarios is ineffective.

**Possible Solutions:**
1. Retry on failed apply with a backoff.
2. Wait for establishment when encountering a CRD. This requires parsing the yaml contents and isn't straightforward for Webhooks, since the definition is different from the actual deployment.

I recommend going with (1) for now, as the easiest thing to implement immediately.

/cc @kkasravi @jlewi @krishnadurai @swiftdiaries @gabrielwen

**Comment by krishnadurai** [comment]
I would like to add that I'm facing:

>     2. Admission Webhooks haven't become ready yet and CR creation fails.

While trying to create ClusterIssuer (a custom resource) for CertManager's deployment since the deployment for it is before the ValidationWebhookConfiguration is deployed.

Refer issue: kubeflow/manifests#543

**Comment by krishnadurai** [comment]
@yanniszark looks like the exponential backoff does not completely resolve this issue.
I'm still running into this error:

```
Error from server (InternalError): error when creating "/tmp/kout": Internal error occurred: failed calling webhook "webhook.cert-manager.io": the server is currently unable to handle the request
```

While deploying cert-manager. Looking into how we can solve this.

**Comment by yanniszark** [comment]
@krishnadurai it looks like the Retry functionality is not working.
It seems that kfctl uses os.Exit() when encountering an error.
I guess that's to be expected, since it's a CLI tool.

From comments in the code, I've seen that this behaviour can also be changed to a panic.
This could work for us. For example, the apply.Apply() function could recover the panic and return an error instead.
Will look into it soon.

**Comment by krishnadurai** [comment]
@yanniszark I concur with you on these findings. apply.Run() seems to be panicking without returning an error.

**Comment by jlewi** [comment]
@yanniszark per discussion in slack; can we get a fix in by EOD today or tomorrow so that we can aim to finalize the release this week?

**Comment by jlewi** [comment]
@yanniszark It looks like the fix has been merged. Is there any way for us to verify this is working with an RC?

**Comment by krishnadurai** [comment]
@jlewi one of the tests to ascertain that webhook establishment is working through retries is to include cert-manager v0.11.0 in KfDef, which earlier faced the same issue as stated here.

**Comment by jlewi** [comment]
@krishnadurai is there a 0.7.0 config which is shipping cert-manager v0.11.0? If not my suggestion would be to close this issue and rely on automatic and manual testing of the different configs that we are shipping for 0.7.0.

**Comment by krishnadurai** [comment]
@jlewi exististing_arrkito config is using cert-manager config.

I will do the verfication to close this issue.

**Comment by krishnadurai** [comment]
The warning messages during retry are missing in the RC: https://github.com/kubeflow/kubeflow/pull/4452

**Comment by jlewi** [comment]
@krishnadurai Does that mean this is fixed and we can close this?

**Comment by krishnadurai** [comment]
This is fixed and we can close this.
